boards: arm: add BeaglePlay on-board CC1352

Ayush1325 opened the pull request:
Add basic support for BeaglePlay, a TI AM6254 SoC based development board with a CC1352P7 wireless microcontroller for supporting software defined 2.4GHz and SubGHz wireless protocols. Support for running Zephyr on the quad-A53 SoC or the programmable M4 on the SoC would be provided separately.

Tested hello world and zperf.

See https://beagleplay.org/ for details.

doc preview: https://builds.zephyrproject.io/zephyr/pr/64718/docs/boards/beagle/beagleplay/doc/beagleplay_cc1352p7.html

jadonk committed:
boards: beagle: add BeaglePlay on-board CC1352P7

Add support for BeaglePlay, a TI AM6254 SoC based development board with a
CC1352P7 wireless microcontroller for supporting software defined 2.4GHz
and SubGHz wireless protocols. Support for running Zephyr on the
quad-A53 SoC or the programmable M4 on the SoC would be provided
separately.

See https://beagleplay.org for details.

Signed-off-by: Jason Kridner <jkridner@beagleboard.org>
Signed-off-by: Ayush Singh <ayush@beagleboard.org>

vaishnavachath commented on boards/arm/beagle_play_cc1352/Kconfig.defconfig:
```diff
@@ -25,15 +25,4 @@ config PINCTRL_CC13XX_CC26XX
-CONFIG_IEEE802154_CC13XX_CC26XX=y
-CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ=y
-CONFIG_WPANUSB_DEVICE_VALID_CHANNELS=0x07FFFFFF
-
```
You are cleaning up a file added in this PR, please squash the cleanup patch into the previous as it is not relevant to git history. Please add the CAPARRAY DELTA changes as part of this PR or https://github.com/zephyrproject-rtos/zephyr/pull/65048

Ayush1325 replied:
done

cfriedt reviewed (dismissed):
LGTM

carlescufi commented:
@vaishnavachath PTAL

vaishnavachath commented on boards/arm/beagle_play_cc1352/board.cmake:
```diff
@@ -0,0 +1,11 @@
+# Be sure to disable the bcfserial driver because it will capture /dev/ttyS4
+
+board_set_flasher_ifnset(misc-flasher)
+board_finalize_runner_args(misc-flasher "$ENV{ZEPHYR_BASE}/boards/arm/beagle_play_cc1352/cc1352-flasher.py --play")
```
@Ayush1325 I still don't see this resolved yet, can you confirm if not having a flasher present in tree is a conscious decision or was it missed in this PR?

Ayush1325 replied:
Sorry, I'm not sure why I didn't see this till now. It is the same as bcf for now. The reason was that the latest script was broken for beagleplay.

Ayush1325 replied:
@vaishnavachath Can you take a look at this: https://github.com/zephyrproject-rtos/zephyr/pull/67229

vaishnavachath commented:
> @vaishnavachath PTAL

@carlescufi it seems like the misc-flasher changes are incomplete, I am not sure how an user would flash to this target without reading the instructions at board.cmake and fixing up the flashing script manually. Other that no concerns with this PR.

vaishnavachath commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,161 @@
+the customized BSL Python script.
+
+* https://docs.beagleboard.org/latest/boards/beagleplay/demos-and-tutorials/zephyr-cc1352-development.html
+
```
Can't you make `west flash` work for this target upstream? I think just using the disable bcfserial overlay and having the flashing script should fix it. Pointing to the above documentation is not really useful for someone trying to develop upstream. Those instructions are specific to downstream repo/there is no specific instructions for independently flashing.

```
Make sure you are in Zephyr directory
cd $HOME/bcf-zephyr (why should someone have a copy of downstream SDK to develop/flash upstream?)
Flash Blinky
cc2538-bsl.py build/blinky
```

@Ayush1325 , please fix this

Ayush1325 replied:
So, the bcfserial overlay would need to be loaded through uboot which means a restart is required. This also would need flasher script whose exclusion was a conscious decision.
I think there are ways to apply overlays dynamically, but haven't tried them. 
I am working on a greybus based solution for flashing and updating firmware.

vaishnavachath replied:
Restart required to flash the device should have nothing to do with having a reliable way to flash the device. A board submitted upstream should have reliable way to build and flash otherwise it is not much useful to other developers who start development.

vaishnavachath replied:
Glad to hear about the effort on greybus based FW update mechanism, that would be nice, but won't those be for the remote nodes/BCF. I remember porting MCUBoot for the CC1352 and using mcumgr for updates (but again reboot will be a concern)

Ayush1325 commented:
@vaishnavachath Can we tackle the question of adding cc1352-flasher in a separate PR. I have updated the docs to install the flasher out of tree for now since #67229 was blocked.

I will be working with upstream [cc2538-bsl](https://github.com/JelmerT/cc2538-bsl) to make the flashing situation better but it will take a while since I do not control those projects.

kartben commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,179 @@
+Debugging
+=========
+
+For debugging, you can use the serial port or JTAG. You can use OpenOCD
+over the Tag-Connect header on the board.
```
are you able to provide openocd.cfg file like many other boards do?

Ayush1325 replied:
No, I haven't pesonally used tag-connect with this board and couldn't find anyone else on discord who has. Maybe @jadonk can provide some input.

Ayush1325 replied:
Ok, so it does not seem like we can provide `openocd.cfg` yet (since no one seems to have used it for cc1352). I will try to get openocd working when I have time.

kartben replied:
If it's never been tested, I think you should be dropping the mention to openOCD for now, no?

vaishnavachath replied:
@Ayush1325 , the openocd configs have been in use for a long time and is the default way to load/flash on a CC135X launchpad.
https://sourceforge.net/p/openocd/code/ci/master/tree/tcl/board/ti_cc13x2_launchpad.cfg
https://sourceforge.net/p/openocd/code/ci/master/tree/tcl/target/ti_cc13x2.cfg

Tagconnect is just a connector, you should be able to reuse the same configuration with Tagconnect.

Ayush1325 replied:
@vaishnavachath How does this look? https://github.com/zephyrproject-rtos/zephyr/pull/64718/commits/762f362ffaa57677c76c1d45beccba2d2b91a524
I don't have the jumper cables to test it myself rn.

jadonk replied:
@vaishnavachath does this work?

kartben commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,179 @@
+
+   .. code-block:: console
+
+      sudo apt install python3-pip
+      pip3 install cc1352-flasher
```
Someone with a working Zephyr dev environment should really have pip already :) 

```suggestion
      pip install cc1352-flasher
```

kartben commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,179 @@
+BeagleBoard.org BeaglePlay is an open hardware single board computer based on a TI Sitara AM6254
+quad-core ARM Cortex-A53 SoC with an external TI SimpleLink multi-standard CC1352P7 wireless MCU
+providing long-range, low-power connectivity.
+
```
can you provide an image? It's always nice to have while not 100% mandatory. Bonus point for having it in webp format, and possibly with transparent background. See ex. https://docs.zephyrproject.org/latest/boards/arm/beagle_bcf/doc/index.html

kartben commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,179 @@
+For debugging, you can use the serial port or JTAG. You can use OpenOCD
+over the Tag-Connect header on the board.
+
+* https://docs.beagleboard.org/latest/accessories/cables.html#tagconnect-jtag
+
+References
+**********
+
+BeagleBoard.org BeaglePlay reference:
+  https://beagleplay.org
```
please check how to make these proper "references" , see other boards like ex. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/bl5340_dvk/doc/index.rst?plain=1

kartben reviewed (changes requested):
this is nice! Some feedback on the doc -- thanks!

jadonk commented on boards/arm/beagle_play_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,187 @@
+For debugging, you can use the serial port or JTAG. You can use OpenOCD
+over the Tag-Connect header on the board.
+
+* https://docs.beagleboard.org/latest/accessories/cables.html#tagconnect-jtag
```
@Ayush1325 can you make this a reference as well?

jadonk commented:
@kartben can we get a fresh review after the updates?

nordicjm reviewed (changes requested):
Needs to be migrated to hwmv2

nordicjm reviewed (changes requested):
hwmv2 changes needs to be applied in initial commit

nordicjm commented on boards/beagle/beagleplay_cc1352/board_antenna.c:
```diff
@@ -0,0 +1,107 @@
+#include <driverlib/rom.h>
+
+/* DIOs for RF antenna paths */
+#define BOARD_RF_HIGH_PA 29 /* TODO: pull from DT */
+#define BOARD_RF_SUB1GHZ 30 /* TODO: pull from DT */
```
@gmarull is it blocking that these are not in dts?

Ayush1325 replied:
Well, this is the exact same file as [beagleconnect_freedom/board_antenna.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/beagle/beagleconnect_freedom/board_antenna.c).

gmarull replied:
I guess it was acceptable to allow this once, but if no proper maintainership is offered (ie, do the to-dos), then -1 on introducing this again.

cfriedt replied:
@Ayush1325 - you can probably just pull in the source from the other directory using a relative path in cmake.

However, I would also like to see an issue (enhancement? Bug maybe?) filed and for @gmarrull's concerns to be addressed from the original source.

@gmarrull - please correct me if I'm wrong, but I think some constants need to be declared in a header file under `include/zephyr/dt-bindings` and for those constants to be incorporated into the board `.dts` or perhaps a parent `.dtsi`. Likely the `.c` file which handles this needs to be cleaned up.

It might help to be very specific about change requests because it might not be obvious what needs to be fixed. An issue would be a good place to be very specific about those things and it can be assigned the a board's owner.

gmarull reviewed (changes requested):
this is misusing gpio-leds, and is using board-level hacks

Ayush1325 commented:
> this is misusing gpio-leds, and is using board-level hacks

You mean `rf_sw` node? That is also the same as what is being used in [beagleconnect_freedom/beagleconnect_freedom.dts](https://github.com/zephyrproject-rtos/zephyr/blob/25812289779f471143cf1a2cc34c245c603bd941/boards/beagle/beagleconnect_freedom/beagleconnect_freedom.dts#L53).

gmarull commented:
> > this is misusing gpio-leds, and is using board-level hacks
> 
> Do you mean `rf_sw` node? That is also the same as what is being used in [beagleconnect_freedom/beagleconnect_freedom.dts](https://github.com/zephyrproject-rtos/zephyr/blob/25812289779f471143cf1a2cc34c245c603bd941/boards/beagle/beagleconnect_freedom/beagleconnect_freedom.dts#L53).
> 
> If these need to be changed, should I start with fixing beagleconnect_freedom (which is already in upstream) or just fix it for play?

It was likely merged without proper reviews. DT is de-facto unmaintained now.

Ayush1325 commented:
@gmarull @nordicjm @cfriedt I have opened https://github.com/zephyrproject-rtos/zephyr/issues/71285 so feel free to add what should be changed in beagleconnect freedom dt there.

I have switched to using `beagleconnect_freedom/board_antenna.c` in this PR. Additionally I will stop the use of `gpio-leds` once I figure out where it `rf_sw` should be placed.

cfriedt commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352.dts:
```diff
@@ -0,0 +1,81 @@
+			label = "LED2";
+		};
+
+		/* U.FL connector switch */
+		rf_sw: rf_sw {
+			label = "RF_SW";
+			gpios =
+				<&gpio0 29 GPIO_ACTIVE_HIGH>, // SubG TX +20dB
+				<&gpio0 30 GPIO_ACTIVE_HIGH>; // SubG TX/RX 0dB
+		};
```
I guess we need some kind of gpio switch bindings in devicetree that behave similarly to the led driver.

@Ayush1325 - to simplify things, group rf_sw inside of its own scope, similarly to the leds, and duplicate the driver.

That would be the easiest solution.

cfriedt replied:
Nice catch 😁

Ayush1325 replied:
I have created the PR to fix this for bcf first: https://github.com/zephyrproject-rtos/zephyr/pull/64718 . Will make changes here once bcf stuff is merged.

jadonk replied:
@Ayush1325 looks like to shared the link to *this* PR, not the one you meant.

jadonk replied:
Correct PR to reference is https://github.com/zephyrproject-rtos/zephyr/pull/71403

Ayush1325 replied:
Fixed: https://github.com/zephyrproject-rtos/zephyr/pull/73488

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

jadonk commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,193 @@
+
+   .. code-block:: console
+
+      echo "    fdtoverlays /overlays/k3-am625-beagleplay-bcfserial-no-firmware.dtbo" | sudo tee -a /boot/firmware/extlinux/extlinux.conf
```
I think this gets the idea across, but it will only work if the booting entry is the last one. Unfortunately, I don't have a good alternate recommendation one-liner as the default extlinux.conf has gotten more complex. There might even be a menu option to select it on boot over serial.

Perhaps just suggesting to load this overlay is what is needed? I'm not against keeping this as-is.

Ayush1325 replied:
Well, we currently need to do different things depending on the kernel version since the old bcfserial overlay does not work in 6.x series. So I think it might be best to remove the overlay from the docs.

jadonk commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,193 @@
+
+   .. code-block:: console
+
+      pip install cc1352-flasher
```
Note: on recent images, this should already be installed. It would be good to make sure it is.

jadonk commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,193 @@
++-------+--------------+-------------------------------------+
+| DIO30 | RF_SUB1G     | Antenna mux SubG enable             |
++-------+--------------+-------------------------------------+
+
```
Note that the antenna mux is only used for antenna J9, the SubG antenna, and muxes between the SubG RX/TX and PA TX paths. J12 is always connected to the 2.4GHz RX/TX path. Internally, I believe it might be possible to use the PA TX path for 2.4GHz, but the design intention is to use J9 for SubG RX and PA TX for SubG signals.

There might be a clearer way to explain that, but as I read the notes, the antenna mux almost implies it is the only antenna.

For simplicity, maybe just call it the "SubG/PA Antenna Mux"?

jadonk commented on boards/beagle/beagleplay_cc1352/support/openocd.cfg:
```diff
@@ -0,0 +1,8 @@
+#
+# TI CC13x2 LaunchPad Evaluation Kit
+#
+# Since bcf does not have a built in debugger, we use XDS110 in launchpad
```
I suggest using "BeagleConnect Freedom", rather than "bcf". Also, "LaunchPad", rather than "launchpad".

jadonk replied:
Er, I even forgot this isn't even "BeagleConnect Freedom", it is "BeaglePlay". I suggest "BeaglePlay CC1352" here.

jadonk commented on boards/beagle/beagleplay_cc1352/beagleplay-cc1352-pinctrl.dtsi:
```diff
@@ -0,0 +1,38 @@
+		input-enable;
+	};
+
+	/* On-board antenna pinmux states */
```
The antenna isn't "On-board". This pinmux feeds J9, where an external SubG antenna provided with BeaglePlay should be connected.

jadonk reviewed (approved):
Thanks for making the antenna switch updates. It looks good. I have a few suggestions for improving the docs.

cfriedt reviewed (commented):
This is looking really nice

jadonk commented:
@gmarull and @nordicjm now that 3.7 is out, can we get updated reviews and wrap this up?

kartben commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352.yaml:
```diff
@@ -0,0 +1,14 @@
+  - gnuarmemb
+  - xtools
+supported:
+  - gpio
```
please complete this list to increase test coverage, ex. `rtc`, `netif:openthread`, `hwinfo`, ...

Ayush1325 replied:
I have added hwinfo since it is present for beagleconnect_freedom as well. Not sure what else should be added. Are all the possible values documented somewhere? I have not tested openthread so that probably should not be added.

gmarull commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352_defconfig:
```diff
@@ -0,0 +1,23 @@
+#
+
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_DYNAMIC=y
+CONFIG_PINCTRL_CC13XX_CC26XX=y
```
this is wrong, where is dynamic used? why is a driver impl enabled? why is pinctrl even enabled here?

Ayush1325 replied:
I have removed dynamic and drive impl option. Pinctrl is required for board antenna init.

gmarull replied:
pinctrl needs to be selected by drivers, not enabled manually like this.

Ayush1325 replied:
So, the board_antenna is not a driver but platform init stuff. So where should I enable PINCTRL? Currently, `beagleconnect_freedom`, `cc1352p1_launchxl` also enable pinctrl in defconfig.

Ayush1325 replied:
I have only left `CONFIG_PINCTRL=y` in the defconfig like most other current boards seem to do and removed `CONFIG_PINCTRL_DYNAMIC=y` and `CONFIG_PINCTRL_CC13XX_CC26XX=y`. Feel free to tell if `CONFIG-PINCTRL` is not supposed to go here in new boards.

cfriedt commented:
@ayush - can you please add a doc preview to your description?

Ayush1325 commented:
> @ayush - can you please add a doc preview to your description?

I am not sure how to do that. Can you point me to an example?

kartben commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+To flash, disable the existing driver that ties up the serial port and use
+the customized BSL Python script.
+
+1. Ensure the bcfserial or gb-beagleplay driver isn’t blocking the serial port. This can be done by
+   loading `/overlays/k3-am625-beagleplay-bcfserial-no-firmware.dtbo` or selecting uboot entry which
+   disables bcfserial/gb-beagleplay.
```
```suggestion
1. Ensure the bcfserial or gb-beagleplay driver isn’t blocking the serial port. This can be done by
   loading :file:`/overlays/k3-am625-beagleplay-bcfserial-no-firmware.dtbo` or selecting uboot entry
   which disables bcfserial/gb-beagleplay.
```

kartben commented:
> > @ayush - can you please add a doc preview to your description?
> 
> I am not sure how to do that. Can you point me to an example?

Done - and spotted a minor formatting issue in the process

@Ayush1325 you can get a link to the CI-built version of the documentation in the results of the Doc build CI job https://github.com/zephyrproject-rtos/zephyr/actions/runs/10279075417

nordicjm commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352.dts:
```diff
@@ -0,0 +1,95 @@
+	 * Path         DIO29 DIO30
+	 * ===========  ===== =====
+	 * Off          0     0
+	 * Sub-1 GHz    0     1     // DIO30 mux to IOC_PORT_RFC_GPO0 for auto
+	 * 20 dBm TX    1     0     // DIO29 mux to IOC_PORT_RFC_GPO3 for auto
+	 */
+	antenna_mux0: antenna_mux0 {
+		compatible = "skyworks,sky13317";
+		status = "okay";
+		gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>, <&gpio0 30 GPIO_ACTIVE_HIGH>;
+		pinctrl-0 = <&board_ant_tx_pa_off &board_ant_subg_off>;
+		pinctrl-1 = <&board_ant_tx_pa_off &board_ant_subg_on>;
+		pinctrl-2 = <&board_ant_tx_pa_on &board_ant_subg_on>;
```
am I misreading this or does the table and pinctrl values not match?

Ayush1325 replied:
So the table here (and in bcf and launchxl) are for GPIO states and not pinctrl. See: https://github.com/zephyrproject-rtos/zephyr/pull/73488#discussion_r1667030803

nordicjm commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_PINCTRL=y
+
+CONFIG_BUILD_OUTPUT_HEX=y
+# custom callback for the antenna switch configuration
```
```suggestion
# Custom callback for the antenna switch configuration
```

nordicjm commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+Supported Features
+==================
+
+The board configuration supports the following hardware features:
```
*board target and put the name you use with west/cmake with \`\` before and after it

nordicjm commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+Supported Features
+==================
+
+The ``beagleconnect_freedom`` board target supports the following hardware features:
```
?? this should be ``beagleplay/cc1352``

Ayush1325 replied:
Right, my mistake. Although I think it should be `beagleplay_cc1352`? That is what is used to build.

Ayush1325 commented on boards/beagle/beagleplay_cc1352/beagleplay_cc1352_defconfig:
```diff
@@ -0,0 +1,21 @@
+#
+
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
```
Well, just doing a basic search, I can see that it is present in a lot of boards:
```
 rg "CONFIG_PINCTRL=y" boards -l | wc -l
296
```
But maybe it is not supposed to go there now? I can see some boards have it defined in the soc `Kconfig.defconfig` so maybe it is supposed to go there now?

gmarull replied:
They do not follow best practices, need to be updated.

Ayush1325 replied:
> They do not follow best practices, need to be updated.

Ok, so where should Pinctrl enabled now?

gmarull replied:
grep for `select PINCTRL` to see patterns

Ayush1325 replied:
> grep for `select PINCTRL` to see patterns

I can see that `espressif` soc do it in `Kconfig` and some drivers seem to do it, I guess. I think you are saying it should be done how `espressif` is doing? Since it is the only one using it, I am not quite sure.

Ayush1325 commented on boards/beagle/beagleplay_cc1352/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
++-----------+------------+----------------------+
+| UART      | on-chip    | serial               |
++-----------+------------+----------------------+
+
```
I will update beagleconnect freedom and other boards separately. For beagleconnect freedom, I also need to add ADC, PWM.

cfriedt reviewed (approved):
LGTM - I think the comments from @vaishnavachath were addressed, but please wait for his approval before merging.

Ayush1325 commented:
> LGTM - I think the comments from @vaishnavachath were addressed, but please wait for his approval before merging.

Sure. But the only change since his approval was a rebase due to a conflict. So I don't think there should be much problem.

jadonk reviewed (approved):
Looks good!

jadonk commented:
Where are the requested changes from @gmarull ?

Ayush1325 commented:
> Where are the requested changes from @gmarull ?

Moved pinctrl to soc config. See `soc/ti/simplelink/cc13x2x7_cc26x2x7/Kconfig`

mmahadevan108 commented:
@Ayush1325 , have you addressed the comments from @gmarull

Ayush1325 commented:
> @Ayush1325 , have you addressed the comments from @gmarull

Maybe my [previous message](https://github.com/zephyrproject-rtos/zephyr/pull/64718#issuecomment-2371808223) was not clear enough. I am assuming the comment everyone is referencing is the following: https://github.com/zephyrproject-rtos/zephyr/pull/64718#discussion_r1703741219

As I stated earlier, `boards/beagle/beagleplay/beagleplay_cc1352p7_defconfig` does not contain any config options regarding pinctrl. I removed `CONFIG_PINCTRL_DYNAMIC=y` and `CONFIG_PINCTRL_CC13XX_CC26XX=y` since they are automatically selected and moved `CONFIG_PINCTRL=y` to `soc/ti/simplelink/cc13x2x7_cc26x2x7/Kconfig` similar to what is being done for `espressif`.

It just seems like github does not automatically dismiss the review, but if you click the "See review option", it does point to a resolved comment (because well, I have addressed everything raised in that comment to the best of my knowledge).

mmahadevan108 commented:
@gmarull, can you please revisit this PR.
